apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java

```diff
@@ -0,0 +1,396 @@
+   *
+   * @param mcsClaim The claim to check.
+   * @return a list of TagDetail objects, one for each SAMHSA code found in the claim.
+   */
+  private Optional<List<TagDetails>> getPossibleMcsSamhsaFields(RdaMcsClaim mcsClaim) {
+    List<TagDetails> entries = new ArrayList<>();
+    LocalDate serviceDate =
+        mcsClaim.getIdrHdrFromDateOfSvc() == null
+            ? LocalDate.parse("1970-01-01")
+            : mcsClaim.getIdrHdrFromDateOfSvc();
+    LocalDate throughDate =
+        mcsClaim.getIdrHdrToDateOfSvc() == null ? LocalDate.now() : mcsClaim.getIdrHdrToDateOfSvc();
+    for (RdaMcsDiagnosisCode diagCode : mcsClaim.getDiagCodes()) {
+      buildDetails(
+          getSamhsaCode(Optional.ofNullable(diagCode.getIdrDiagCode())),
+          "mcs_diagnosis_codes",
+          "idr_diag_code",
+          (int) diagCode.getRdaPosition(),
+          entries,
+          serviceDate,
+          throughDate);
+    }
+    for (RdaMcsDetail detail : mcsClaim.getDetails()) {
+      buildDetails(
+          getSamhsaCode(Optional.ofNullable(detail.getIdrDtlPrimaryDiagCode())),
+          "mcs_details",
+          "idr_dtl_primary_diag_code",
+          (int) detail.getIdrDtlNumber(),
+          entries,
+          serviceDate,
+          throughDate);
+
+      buildDetails(
+          getSamhsaCode(Optional.ofNullable(detail.getIdrProcCode())),
+          "mcs_details",
+          "idr_proc_code",
+          (int) detail.getIdrDtlNumber(),
+          entries,
+          serviceDate,
+          throughDate);
+    }
+    return entries.isEmpty() ? Optional.empty() : Optional.of(entries);
+  }
+
+  /**
+   * Builds a TagDetails object, and adds it to a list of TagDetails.
+   *
+   * @param entry The SamhsaEntry that holds information about the SAMHSA code.
+   * @param table The table that the code belongs to
+   * @param column The column that the cod belongs to
+   * @param lineNum The line number that the code belongs to
+   * @param detailsList The TagDetails list to add to
+   * @param serviceDate The first date of service for the claim
+   * @param throughDate The service through date
+   */
+  private void buildDetails(
+      Optional<SamhsaEntry> entry,
+      String table,
+      String column,
+      Integer lineNum,
+      List<TagDetails> detailsList,
+      LocalDate serviceDate,
+      LocalDate throughDate) {
+    if (entry.isPresent()) {
+      try {
+        LocalDate startDate = LocalDate.parse(entry.get().getStartDate());
+        LocalDate endDate =
+            entry.get().getEndDate().equalsIgnoreCase("Active")
+                ? LocalDate.now()
+                : LocalDate.parse(entry.get().getEndDate());
+
+        // if the throughDate is not between the start and end date,
+        // and the serviceDate is not between the start and end date,
+        // then the claim falls outside the date range of the SAMHSA code.
+        if ((throughDate.isBefore(startDate) || throughDate.isAfter(endDate))
```

Comment:
This looks correct, but it's a bit hard to follow. Maybe it would help with readability to make an `isBetween` helper function?